net/icmplistener/icmplistener.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build linux || darwin
+
+// Package icmplistener implements a net.ListenConfig interface that overrides
+// the handling of "ip:icmp" and "ip6:icmp" networks to use datagram sockets
+// instead of raw sockets.
+//
+// In the 2000s the prevalence of ICMP based internet attacks led to broad
+// consensus that raw sockets must be highly priveleged, causing all ICMP to
+// become unavailable to unprivileged processes. In more recent years, standing
+// concerns about extending privelege to keep `ping` working have lead to a new
+// emerging consensus that ICMP Echo specifically should be allowed, and the
+// mechanism for doing so is to send ICMP Echo packets via a SOCK_DGRAM socket
+// type.
+//
+// This behavior is implemented by macOS and Linux (in Linux this is contingent
+// on `net.ipv4.ping_group_range` covering the users range, which it typically
+// does).
+//
+// The Go net abstraction does not directly lend itself to this kind of
+// reimplementation, as such some edge case behaviors may differ in deliberately
+// undocumented ways. Those behaviors may later change to fit intended use cases
+// (initially sending ICMP Echo from userspace).
+package icmplistener
+
+import (
+	"context"
+	"net"
+	"net/netip"
+	"os"
+
+	"golang.org/x/sys/unix"
+)
+
+type ListenConfig struct {
+	net.ListenConfig
+}
+
+func (lc *ListenConfig) ListenPacket(ctx context.Context, network, address string) (net.PacketConn, error) {
+	switch network {
+	case "ip:icmp", "ip6:icmp", "ip4:icmp", "ip6:icmp-ipv6":
+		return lc.listenICMP(ctx, network, address)
```

Comment:
would need docs on exported stuff (perhaps moving some of the package doc)

But what's the purpose of this magic here? Is there somewhere we need to plumb a ListenConfig-like thing that does both ICMP pings and UDP? Can this package be ping-specific instead? But then if so, what's the relationship between this package and our net/ping.Pinger stuff?

Reply:
the plan was to integrate this into netns, so that the socketmark infrastructure doesn't need new plumbing to more places, currently the netns listenconfig is what's passed in to the net/ping constructor